src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -1546,6 +1631,79 @@ protected override void HandleThreadsRequestAsync(IRequestResponder<ThreadsArgum
+                var version = debugModuleInfo.m_bstrVersion;
+                var vsLoadAddress = debugModuleInfo.m_addrLoadAddress.ToString(CultureInfo.InvariantCulture);
+                var vsPreferredLoadAddress = debugModuleInfo.m_addrPreferredLoadAddress.ToString(CultureInfo.InvariantCulture);
+                var vsModuleSize = (int)debugModuleInfo.m_dwSize;
+                var vsLoadOrder = (int)debugModuleInfo.m_dwLoadOrder;
+                var symbolFilePath = debugModuleInfo.m_bstrUrlSymbolLocation;
+                var symbolStatus = debugModuleInfo.m_bstrDebugMessage;
+                var vsIs64Bit = (debugModuleInfo.m_dwModuleFlags & enum_MODULE_FLAGS.MODULE_FLAG_64BIT) != 0;
+
+                ProtocolMessages.Module mod = new ProtocolMessages.Module(moduleId, debugModuleInfo.m_bstrName)
+                {
+                    Path = path, VsTimestampUTC = vsTimestampUTC, Version = version, VsLoadAddress = vsLoadAddress, VsPreferredLoadAddress = vsPreferredLoadAddress,
+                    VsModuleSize = vsModuleSize, VsLoadOrder = vsLoadOrder, SymbolFilePath = symbolFilePath, SymbolStatus = symbolStatus, VsIs64Bit = vsIs64Bit
+                };
+            
+                // IsOptimized and IsUserCode are not set by gdb
+                if((debugModuleInfo.m_dwModuleFlags & enum_MODULE_FLAGS.MODULE_FLAG_OPTIMIZED) != 0)
+                {
+                    mod.IsOptimized = true;
+                } else if ((debugModuleInfo.m_dwModuleFlags & enum_MODULE_FLAGS.MODULE_FLAG_UNOPTIMIZED) != 0)
+                {
+                    mod.IsOptimized = false;
+                }
+                if (module is IDebugModule3 module3 && module3.IsUserCode(out int isUserCode) == HRConstants.S_OK)
+                {
+                    if (isUserCode == 0)
+                    {
+                        mod.IsUserCode = false;
+                    }
+                    else
+                    {
+                        mod.IsUserCode = true;
+                    }
+                }
+
+                return mod;
+            }
+            return null;
+        }
+
+        protected override void HandleModulesRequestAsync(IRequestResponder<ModulesArguments, ModulesResponse> responder)
+        {
+            var response = new ModulesResponse();
+
+            IEnumDebugModules2 enumDebugModules;
+            if (m_program.EnumModules(out enumDebugModules) == HRConstants.S_OK)
+            {
+                var debugModules = new IDebugModule2[1];
+                uint numReturned = 0;
+                while (enumDebugModules.Next(1, debugModules, ref numReturned) == HRConstants.S_OK && numReturned == 1)
+                {
+                    IDebugModule2 module = debugModules[0];
+                    var moduleId = RegisterDebugModule(module);
+                    if (moduleId != null)
```

Comment:
You don't want `RegisterDebugModule` here -- the module should already be in the collection.

```suggestion
                    int moduleId;
                    lock (m_moduleMap)
                    {
                        if (!m_moduleMap.TryGetValue(debugModule, out moduleId))
                        {
                            Debug.Fail("Missing ModuleLoadEvent?");
                            continue;
                        }
                    }
                    
                    var mod = ConvertToModule(module, moduleId);
                    response.Modules.Add(mod);
```

Reply:
Sorry, blame Andrew. He gave you bad advice :).

Reply:
Andrew, is there any case I am not thinking of where these module could be new? If so, instead of asserting and ignoring, you could call `RegisterModule` there.

Reply:
Yep. I misunderstood how IDebugProgram's module's list and IDebugModuleEvent worked. I got it clarified.

They should all be existing, so asserting is fine.

Reply:
@gregg-miskelly Do we want to support ModulesArguments's `startModule` and `moduleCount`? Or do we want to always return the full module list.

`startModule` is going to be weird with the Dictionary implementation of keeping track of the modules.

Reply:
In looks like VS will always request all modules. I am not sure why any IDE would want a subset. Probably fine to just return all of them until we have a reason to implement it.